### ISSUE TITLE
Backbone-Metal

### ISSUE THREAD

**Issue by paulfalgout** [marionettejs/backbone.marionette#2426]
:metal: 

As the champion of Metal has moved on to greener pastures,  we need to take a hard look at Metal, understand what it means for Marionette and figure out how to proceed.  @marionettejs/marionette-core 

https://github.com/marionettejs/backbone-metal seems to be inspired by http://emberjs.com/api/modules/ember-metal.html

The purpose of Ember-metal is not completely clear http://numberofpeoplewhounderstandembermetal.com/

So what does Backbone-Metal do, and what is its purpose in regards to Marionette?  I'll attempt a first pass at some explanations.  No doubt I'll misrepresent something.. so please chime in.

Very generally:
* It adds a generic `Class` to replace Marionette's `Object` formerly known as `Controller`.  
* It adds `Mixin` functionality which is very similar to `extend` but it promotes a pattern of sharing functionality vs deep inheritance. 
* It wraps each function so that you can call _super which is a generalized way to call the same function from the inherited Class.

In general I support these features, but some cons might be:
* `Marionette.Class` and `Marionette.Object` are not all that different in providing a generic base element to extend from.  Is there really that much benefit in renaming it yet again?
* The suggested implementation of metal on Marionette patches Backbone as well for `_super` to work with Backbone classes.
* ~~To get `_super` to work, _every_ function _everywhere_ must be wrapped.~~  Are there expensive implications to this extra wrap and step in the scope chain?

Regarding the general purpose of extracting the generic object from Marionette as well as errors, deprecation and other helpers, I believe this is the first step in making Marionette's code specifically more concerned with Views alone.  I am on the fence as far as the necessity of making this a stand-alone library.  The fact that it is also ES6, in the current javascript climate, I believe, really makes it less accessible to the general population of developers.  And I am :-1: on making Marionette more difficult to understand from soup to nuts.  I am also :-1: on having a separate library [pick](https://github.com/marionettejs/backbone-metal/issues/3) and [choose](https://github.com/marionettejs/backbone-metal/issues/4) what base Marionette features to support.

Thoughts?

**Comment by capricube** [comment]
Where did the champion go to?
Am 17.03.2015 07:15 schrieb "Paul Falgout" <notifications@github.com>:

**Comment by jasonLaster** [comment]
Thanks for the writeup. I'll share more thoughts later, but perf wise I believe we're only wrapping the functions who override ancestor functions. It's slower, but not too much.

**Comment by cmaher** [comment]
So https://github.com/lukasolson/backbone-super is already a thing, and it provides opt-in `_super` wrapping, rather than forcing it. To get it working with marionette, a small shim is needed:

```js
    Marionette.extend = Backbone.Model.extend;
    Marionette.View.extend = Marionette.extend;
    Marionette.ItemView.extend = Marionette.extend;
    Marionette.CompositeView.extend = Marionette.extend;
    Marionette.Region.extend = Marionette.extend;
    Marionette.RegionManager.extend = Marionette.extend;
    Marionette.LayoutView.extend = Marionette.extend;
```
which could be fixed by having Marionette.extend proxy Backbone.Model.extend instead of referencing it directly:

```js
Marionette.extend = function () { 
  return Backbone.Model.extend.apply(this, arguments);
};
```

Mixins can be similarly handled through default Backbone extension

```js
var mixin1 = { stuff: 'woo' };
var mixin2 = { moreStuff: 'wooooooo' };
var MyModel = Backbone.Model.extend(mixin).extend(mixin2).extend({
});
```
which combines nicely with backbone.super.

In short, we don't need Metal. The Backbone community is very much a choose-your-own-framework kind of world, which definitely favors small plugins that work well together vs one monolithic framework.  We could do much more for the community if we worked with other plugin authors to make sure everything works nicely together, instead of just rewriting existing plugins and adding it to Marionette directly.

**Comment by paulfalgout** [comment]
@cmaher I think I mostly agree.

However if Mn wants to internally use mixins and `_super` how do you feel about adding another dependency?  

I suppose the question comes down to:  Is `_super` important enough for first class support either by our own library or a 3rd party dependency, or is it only necessary that there's a guide and a cookbook type of thing?

As to the rename of `Object` to `Class` looking back at https://github.com/marionettejs/backbone.marionette/issues/2274  it looks like the main concern is:
```js
import {Object} from 'marionette'; 
```

**Comment by thejameskyle** [comment]
There's a lot more to having a base object that everything else extends other than `_super`. If you want to make Marionette an extensible from a low-level there needs to be some kind of consistency.

For one being able to tell when something is of a certain type:

```js
function(param) {
  if (param instanceof Metal.Class) {
    // ...
  } else if (Metal.Class.isClass(param)) {
    // ...
  } else if (_.isFunction(param)) {
    // ...
  } else if (_.isObject(param)) {
    // ...
  }
}
```

Or to be able to build smarter debugging tools

```js
// after metal and before any other code
Metal.Class = Metal.Class.extend({
  constructor(...args) {
    sendInstanceToInspector(this);
    this._super(...args);
  }
});

class MyClass extends Metal.Class {}
new MyClass(); // sent to inspector
```

I can add basic functionality to all of my Classes:

```js
Metal.Class.prototype.getDeep = function(attr) {
  return _.reduce(attr.split('.'), (memo, name) => {
    return memo && memo[name];
  }, this);
};
```

Try being a bit more imaginative and you'll quickly find hundreds awesome things you can do.

**Comment by thejameskyle** [comment]
> To get _super to work, every function everywhere must be wrapped. Are there expensive implications to this extra wrap and step in the scope chain?

This couldn't be less true, take a second and read the code.

**Comment by thejameskyle** [comment]
Greener pastures indeed...

**Comment by paulfalgout** [comment]
:point_right:  :bear:

:wave: @thejameskyle appreciate the insightful comments.  I certainly do not want to misrepresent the great work there.  Metal ( :metal: ) is potentially a sizable, foundational change to Marionette, and one other issues seem to be waiting on for implementation.  When the original issues/PRs were closed, it seemed the conversation needed to continue without much delay.  I clearly don't know enough about it to defend it entirely on my own, so I took a quick stab at it to float any possible concerns to the top, hoping they'd be refuted or that they'd lead to improvements or at least ignite some sort of continued conversation.

In other news, @cmaher  in an equally hasty review of https://github.com/lukasolson/backbone-super I doubt this library supports IE8.  They have a try{ } finally { } without a catch, which I believe is an IE8 no-no.

**Comment by jasonLaster** [comment]
Another example, that I believe closely aligns with bb metal is Ember core object. The implementation is much more complicated though: https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/system/core_object.js

----

Consolidating a bunch of stuff:
+ Paul: Move from  Object to class is largely so that Class does not collide
+ Perf: Only the functions that call `super` are [wrapped](https://github.com/marionettejs/backbone-metal/blob/master/src/backbone-metal.js#L61-L62)
+ Inheritance: One could add a function to Mn.Class and all subclasses would get it
+ Compatability: instanceof and typeof work `param instanceof Metal.Class`
+ How does BB.Super compare to Mn.Metal to Ember.Metal.CoreObject?
+ Do we want a recipe for `super` or get it out of the box?

----

I'm personally in favor of adding Metal and additional "base" things to Marionette:
1. It'lll help us as framework authors build a better Marionette.
2. It's hard to ask our community to adopt low-level functionality. Most will be queasy if they open the file, if we shoulder the responsibility by making a lib a hard depenedency the community will go along.

In terms of rolling our own or using a community lib, I'd prefer to use a community lib if it works for us because it widens the tent. In this case, after talking with Stephen (Ember CoreObject) and James (BB Metal), it seems like they're pretty philosophically aligned which gives me some comfort. 

In terrms of responsibility, I think we as a group would have to step up and take ownership of metal if we adopt it. I feel like I can follow BB.Metal internals and others on the team can as well. This gives me some comfort as well. I'd be happy to champion it's integration and then make sure we're all ramped up.

**Comment by paulfalgout** [comment]
This is a WIP.  See #2534 #2535  etc
